KAFKA-7597: Make Trogdor ProduceBenchWorker support transactions

stanislavkozlovski opened the pull request:
It now accepts a new "messagesPerTransaction" field, which, if present, will enable transactional functionality in the bench worker.
The producer will open N transactions with X messages each (bounded by the mandatory "maxMessages" field)

stanislavkozlovski committed:
KAFKA-7597: Make Trogdor ProduceBenchWorker support transactions

It now accepts a new "messagesPerTransaction" field, which, if present, will enable transactional functionality in the bench worker.
The producer will open N  transactions with X messages each (bounded by the mandatory "maxMessages" field)

stanislavkozlovski commented:
Retest this please
cc @cmccabe for review

cmccabe commented:
Hmm.  Just having a static "number of messages per transaction" field seems kind of limiting, right?  What if we later want a mix of transactional and non-transactional messages, or transactions of different sizes?

I would suggest making something like `PayloadGenerator`.  Perhaps `TransactionGenerator`?
Then we can have `ZeroTransactionsGenerator` (the default if none is configured) and `UniformSizeTransactionsGenerator`

stanislavkozlovski committed:
Drive produceBench transactions through a new TransactionGenerator field

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/TransactionActionGenerator.java:
```diff
@@ -0,0 +1,47 @@
+    @JsonSubTypes.Type(value = UniformTransactionsGenerator.class, name = "uniform"),
+})
+public interface TransactionActionGenerator {
+    enum TransactionActions {
```
nit: makes more sense as the singular `TransactionAction` than as the plural `TransactionActions` since each value represents a single action

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/UniformTransactionsGenerator.java:
```diff
@@ -0,0 +1,62 @@
+
+    private final int messagesPerTransaction;
+    private boolean initialized = false;
+    private int messagesInTransaction = -1;
```
should be final

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/TransactionActionGenerator.java:
```diff
@@ -0,0 +1,47 @@
+    include = JsonTypeInfo.As.PROPERTY,
+    property = "type")
+@JsonSubTypes(value = {
+    @JsonSubTypes.Type(value = ZeroTransactionsGenerator.class, name = "zeroTransactions"),
```
perhaps "zero" is better for a short name used in configuration?  Although I guess most people will configure this by leaving it blank.

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/UniformTransactionsGenerator.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @Override
+    public TransactionActions nextAction() {
```
It's probably better to make this synchronized, unless we want to say that TransactionActionGenerator classes are not thread-safe

stanislavkozlovski replied:
good catch

cmccabe commented:
Looks good overall.  I left some comments

stanislavkozlovski committed:
Address PR comments

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/TransactionActionGenerator.java:
```diff
@@ -0,0 +1,47 @@
+    @JsonSubTypes.Type(value = ZeroTransactionsGenerator.class, name = "zero"),
+    @JsonSubTypes.Type(value = UniformTransactionsGenerator.class, name = "uniform"),
+})
+public interface TransactionActionGenerator {
```
Sorry to keep talking about the name of this, but I just don't like how `TransactionActionGenerator` sounds.  `TransactionGenerator` sounds much better.

Then we could have `TransactionGenerator#Action` as well and it flows really well

cmccabe commented:
Looks almost ready.  Left some minor comments

stanislavkozlovski committed:
Renames

stanislavkozlovski commented:
An unrelated test failed in JDK 8 -the dreaded `kafka.admin.ReassignPartitionsClusterTest.shouldExecuteThrottledReassignment`. On an unrelated note, I am currently working on fixing that exact test's flakiness

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchWorker.java:
```diff
@@ -179,18 +182,34 @@ protected synchronized void delay(long amount) throws InterruptedException {
+            this.partitionsIterator = activePartitions.iterator();
             this.histogram = new Histogram(5000);
+
+            this.txActionGenerator = spec.transactionGenerator();
```
Can we name this `transactionGenerator` for simplicity?

stanislavkozlovski replied:
yep, good catch on the leftover

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/TransactionGenerator.java:
```diff
@@ -0,0 +1,47 @@
+     * to signal the producer that its next step should be to send a message.
+     *
+     * The first returned action of every generator should be #{@link TransactionAction#INIT_TRANSACTIONS},
+     * in order to signal to the producer that it should be configured for using transactions
```
If we use `Optional<TransactionGenerator>` we can get rid of `INIT_TRANSACTIONS` (it will be implied if the Optional contains a generator)

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchWorker.java:
```diff
@@ -179,18 +182,34 @@ protected synchronized void delay(long amount) throws InterruptedException {
+            this.txActionGenerator = spec.transactionGenerator();
+            this.transactionsCommitted = new AtomicInteger();
+            if (txActionGenerator.action() == TransactionAction.INIT_TRANSACTIONS)
+                toUseTransactions = true;
```
I think we should check to make sure the first action is either `INIT_TRANSACTIONS` or `NO_OP`.  Other values should not be allowed.

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchWorker.java:
```diff
@@ -179,18 +182,34 @@ protected synchronized void delay(long amount) throws InterruptedException {
-            // add common client configs to producer properties, and then user-specified producer
-            // configs
+            if (toUseTransactions)
+                props.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, "transaction-id-" + UUID.randomUUID());
```
The transactional ID should include `produce-bench-` in it somewhere to make it clear what test created this, right?

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchWorker.java:
```diff
@@ -179,18 +182,34 @@ protected synchronized void delay(long amount) throws InterruptedException {
+        private Iterator<TopicPartition> partitionsIterator;
+        private Future<RecordMetadata> sendFuture;
+        private AtomicInteger transactionsCommitted;
+        private boolean toUseTransactions = false;
```
Suggest a name like `enableTransactions`

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/TransactionGenerator.java:
```diff
@@ -0,0 +1,47 @@
+    include = JsonTypeInfo.As.PROPERTY,
+    property = "type")
+@JsonSubTypes(value = {
+    @JsonSubTypes.Type(value = ZeroTransactionsGenerator.class, name = "zero"),
```
I thought about this a little bit more, and I think maybe we should just use `Optional<TransactionGenerator>` rather than having `ZeroTransactionsGenerator`.  The zero generator isn't a real generator and it's kind of confusing

stanislavkozlovski committed:
Rename variable and checkstyle fix

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchWorker.java:
```diff
@@ -273,6 +338,7 @@ StatusData update() {
         private final int p50LatencyMs;
         private final int p95LatencyMs;
         private final int p99LatencyMs;
+        private final int transactionsCommitted ;
```
This is a checkstyle issue.  Are you compiling with `check` to catch these before Jenkins does?

stanislavkozlovski committed:
Remove zeroTransactionGenerator class and make use of Optionals jackson vars

Imports new library - jackson-datatype-jdk8. This should be used until jackson 3 comes out, at which point we can switch to it completely
Removed INIT_TRANSACTIONS state from the `TransactionAction` enum, this eases TransactionGenerator implementations
Removed ZeroTransactionsGenerator class - it is much more intuitive to provide null or not define the transactionsGenerator class

ijuma commented on gradle/dependencies.gradle:
```diff
@@ -103,6 +103,7 @@ libs += [
   bcpkix: "org.bouncycastle:bcpkix-jdk15on:$versions.bcpkix",
   easymock: "org.easymock:easymock:$versions.easymock",
   jacksonDatabind: "com.fasterxml.jackson.core:jackson-databind:$versions.jackson",
+  jacksonDatatype: "com.fasterxml.jackson.datatype:jackson-datatype-jdk8:$versions.jackson",
```
I think we should call this jacksonJdk8 or something.

ijuma commented on build.gradle:
```diff
@@ -918,6 +918,7 @@ project(':tools') {
     compile project(':log4j-appender')
     compile libs.argparse4j
     compile libs.jacksonDatabind
+    compile libs.jacksonDatatype
```
We should probably add this to other projects that depend on `Jackson` so that we get reasonable behaviour for Java 8 types.

stanislavkozlovski committed:
Rename jackson jdk8 lib name and add to all projects that use jackson

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchWorker.java:
```diff
@@ -285,19 +350,26 @@ StatusData update() {
         }
 
+        @JsonProperty
+        public long transactionsCommitted() {
```
Hmm.  This is a long here, but an int in other places.  Should this be a `long` everywhere?

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/TransactionGenerator.java:
```diff
@@ -0,0 +1,43 @@
+     * This means that most of the time it should return #{@link TransactionAction#NO_OP}
+     * to signal the producer that its next step should be to send a message.
+     */
+    TransactionAction action();
```
`nextAction` might be a better name here

stanislavkozlovski committed:
Rename txGenerator#action() method, use long for transactionsCommitted var, rename transactional id

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchSpec.java:
```diff
@@ -67,6 +98,7 @@ public ProduceBenchSpec(@JsonProperty("startMs") long startMs,
             new SequentialPayloadGenerator(4, 0) : keyGenerator;
         this.valueGenerator = valueGenerator == null ?
             new ConstantPayloadGenerator(512, new byte[0]) : valueGenerator;
+        this.transactionGenerator = txGenerator;
```
We should convert null -> none here.  Even if Jackson always fills this in, someone may manually create this object

stanislavkozlovski committed:
Check for null when given transactionGenerator in ProduceBenchSpec

cmccabe commented:
LGTM
